_posts/2023/2023-03-13-assign-a-default-to-an-attribute-active-record.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "Assign a default value to an attribute in Active Record"
+description: "You’ve probably been using callbacks. Don’t."
+layout: article
+category: ruby
+image:
+  base: '2023/assign-default-to-an-attribute-active-record'
+  alt: "Blank wood"
+  credit: "Keith Misner"
+  source: "https://unsplash.com/photos/h0Vxgz5tyXA"
+
+---
+
+If you ever needed to set a default value in an Active Record model, you probably used a callback.
+
+Since Rails 5.0 there’s [been a better way](https://edgeguides.rubyonrails.org/5_0_release_notes.html#active-record-attributes-api). I had missed it until recently (thanks [Moses](https://twitter.com/Gathukumose/status/1615298667031797760)!).
+
+
+## Instead of…
+
+…assigning a default value in a callback:
+
+```ruby
+class Message
+  before_validation :assign_delivered_at
+
+  # ...
+
+  private
+
+  def assign_delivered_at
+    delivered_at ||= Time.zone.now
+  end
+end
+```
+
+
+## Use…
+
+…the attribute API from Active Record:
+
+```ruby
+class Message
+  attribute :delivered_at, default: -> { Time.zone.now }
+
+  # ...
+end
+```
+
+
+## Why?
+
+Callbacks can be confusing to understand even when there's a good reason to use them. Generally the less you use them the less surprises you’ll have at a later date.
+
+This more recent API is terser, clearer and _the way_ Rails recommends to execute this behaviour.
+
+There’s a lot more going on [in the Attributes API](https://api.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html)—Dirty tracking, type casting, attributes not backed by the database—but in this case we’re only using the default-setting capability.
```

Comment:
Oh, "Attributes API" — then that's what it should be everywhere else before!

And then: *There's a lot more going on in this API— change tracking (instead of Dirty for those that don't know? Or is it commonly known? Would it be capitalised?), type casting...

Reply:
```suggestion
There’s a lot more going on [in this API](https://api.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html)—attribute changes tracking, type casting, adding attributes not backed by the database—but in this case we’re only using the default-setting capability.
```